### ISSUE TITLE
Opening/Closing analytic view with graph opened causes RenderExceptions

### ISSUE THREAD

**Issue by aldebaran30701** [constellation-app/constellation#646]
### Prerequisites

* [ ] Put an X between the brackets on this line if you have done all of the following:
    * Running the latest version of Constellation
	* Attached the ***Support Package*** via `Help` > `Support Package`
    * Checked the FAQs: https://github.com/constellation-app/constellation/wiki/FAQ
    * Checked that your issue isn't already filed: https://github.com/constellation-app/constellation/issues
    * Checked that there is not already a module that provides the described functionality: https://github.com/constellation-app/constellation/wiki/Catalogue-of-Repositories

### Description

Trying to recreate #629 I managed to find a rendering exception.
I used the hotkeys ctrl + shift + z to open analytic view, and closed it via the 'x' button on the tab. I repeated this decently fast and 1/10th of the time I would get this exception. I've recreated it twice within a few minutes.

```
au.gov.asd.tac.constellation.visual.opengl.utilities.RenderException: Incorrect amount of data for the specified target [size=3504, limit=5164, position=0]
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.Batch.bufferData(Batch.java:367)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.Batch.buffer(Batch.java:296)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.NodeLabelBatcher.lambda$createBatch$0(NodeLabelBatcher.java:162)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GraphRenderable.lambda$update$29(GraphRenderable.java:378)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GraphRenderable.update(GraphRenderable.java:377)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GLRenderer.lambda$display$2(GLRenderer.java:166)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GLRenderer.display(GLRenderer.java:165)
	at jogamp.opengl.GLDrawableHelper.displayImpl(GLDrawableHelper.java:692)
	at jogamp.opengl.GLDrawableHelper.display(GLDrawableHelper.java:674)
	at com.jogamp.opengl.awt.GLCanvas$11.run(GLCanvas.java:1428)
	at jogamp.opengl.GLDrawableHelper.invokeGLImpl(GLDrawableHelper.java:1293)
Caused: com.jogamp.opengl.GLException: Caught RenderException: Incorrect amount of data for the specified target [size=3504, limit=5164, position=0] on thread AWT-EventQueue-0
	at com.jogamp.opengl.GLException.newGLException(GLException.java:76)
	at jogamp.opengl.GLDrawableHelper.invokeGLImpl(GLDrawableHelper.java:1327)
	at jogamp.opengl.GLDrawableHelper.invokeGL(GLDrawableHelper.java:1147)
	at com.jogamp.opengl.awt.GLCanvas$12.run(GLCanvas.java:1442)
	at com.jogamp.opengl.Threading.invoke(Threading.java:223)
	at com.jogamp.opengl.awt.GLCanvas.display(GLCanvas.java:503)
	at com.jogamp.opengl.awt.GLCanvas.paint(GLCanvas.java:557)
	at java.desktop/sun.awt.RepaintArea.paintComponent(RepaintArea.java:264)
	at java.desktop/sun.awt.RepaintArea.paint(RepaintArea.java:240)
	at java.desktop/sun.awt.windows.WComponentPeer.handleEvent(WComponentPeer.java:358)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5069)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
[catch] at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

```

### Steps to Reproduce

1. open graph and use complete graph builder
2. open new graph and use sphere graph builder
3. ctrl + shift + z to open analytic view 
4. close the view
5. repeat 3 and 4 until exception occurs

**Expected behaviour:** no errors

**Actual behaviour:** render exception

**Reproduces how often:** 10%

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by aldebaran30701** [comment]
After the fix for #629 was merged in, I was able to trigger another exception with rendering.
@Nova-2119 Would you know if the stack trace lines relate to anything done in the fix? It could have just as well been another issue altogether and not as a result of your fix.
```
au.gov.asd.tac.constellation.visual.opengl.utilities.RenderException: Can't initialise once a batch has been finalised on a GL context. Dispose it first.
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.Batch.initialise(Batch.java:207)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.batcher.NodeLabelBatcher.lambda$createBatch$0(NodeLabelBatcher.java:161)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GraphRenderable.lambda$update$29(GraphRenderable.java:378)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GraphRenderable.update(GraphRenderable.java:377)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GLRenderer.lambda$display$2(GLRenderer.java:166)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GLRenderer.display(GLRenderer.java:165)
	at jogamp.opengl.GLDrawableHelper.displayImpl(GLDrawableHelper.java:692)
	at jogamp.opengl.GLDrawableHelper.display(GLDrawableHelper.java:674)
	at com.jogamp.opengl.awt.GLCanvas$11.run(GLCanvas.java:1428)
	at jogamp.opengl.GLDrawableHelper.invokeGLImpl(GLDrawableHelper.java:1293)
Caused: com.jogamp.opengl.GLException: Caught RenderException: Can't initialise once a batch has been finalised on a GL context. Dispose it first. on thread AWT-EventQueue-1
	at com.jogamp.opengl.GLException.newGLException(GLException.java:76)
	at jogamp.opengl.GLDrawableHelper.invokeGLImpl(GLDrawableHelper.java:1327)
	at jogamp.opengl.GLDrawableHelper.invokeGL(GLDrawableHelper.java:1147)
	at com.jogamp.opengl.awt.GLCanvas$12.run(GLCanvas.java:1442)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:303)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused: java.lang.RuntimeException
	at com.jogamp.common.util.awt.AWTEDTExecutor.invoke(AWTEDTExecutor.java:58)
	at jogamp.opengl.awt.AWTThreadingPlugin.invokeOnOpenGLThread(AWTThreadingPlugin.java:103)
	at jogamp.opengl.ThreadingImpl.invokeOnOpenGLThread(ThreadingImpl.java:201)
	at com.jogamp.opengl.Threading.invokeOnOpenGLThread(Threading.java:202)
	at com.jogamp.opengl.Threading.invoke(Threading.java:221)
	at com.jogamp.opengl.awt.GLCanvas.display(GLCanvas.java:503)
	at au.gov.asd.tac.constellation.visual.opengl.renderer.GLVisualProcessor.performVisualUpdate(GLVisualProcessor.java:197)
	at au.gov.asd.tac.constellation.utilities.visual.VisualProcessor.lambda$update$0(VisualProcessor.java:211)
[catch] at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by Nova-2119** [comment]
@aldebaran30701 Sorry it took so long to reply, only just saw this.

I don't believe my fix could have cause this behaviour. The closest it gets is in the NodeLabelBatcher which has a two methods which could lead to the code changed in my fix. But neither of those methods is called in this stack trace (bufferBottomLabel and bufferTopLabel).

Good thinking though.

For what its worth I have also seen a couple of different errors when rendering graphs. Weirdly neither had any affect on Constellation behaviour as far as I could tell...I was able to close them and keep acting as normal. I think the original error in this ticket is one of the ones I have seen. I haven't seen the second one.

**Comment by aldebaran30701** [comment]
@Nova-2119 I appreciate the response, I too did not notice any change in the normal behaviour. My worry is that it could be unnoticed and just not rendering a section of graph.

**Comment by Nova-2119** [comment]
@aldebaran30701 I had the same thought.
I did check for that once and everything seemed to be there, but I'm keeping an eye out for another opportunity to check.

**Comment by aldebaran30701** [comment]
#677 occurs in the V2.0 rc2 testing (and v1.2) instead of this case - Unsure if it is directly related or a different issue altogether.

**Comment by serpens24** [comment]
I'm also able to replicate this. The exception is thrown here: https://github.com/constellation-app/constellation/blob/master/CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/Batch.java
![image](https://user-images.githubusercontent.com/58677108/89248902-005c5400-d650-11ea-93b8-5aefc0e8c35d.png)

After speaking to @aldebaran30701 I found having the mouse hovering over the position of the 'X' on the analytic view and quickly clicking it after the Ctrl_Shift+Z seems to be the best way to trigger problem.

As an exercise I commented out the exception being thrown and replaced with a comment and found the comment is displayed at semi regular intervals  ... I didnt notice anyhting adverse occurring.... however if worried I found that inserting a return call as well means the subsequent code in the function is skipped (although this doesnt handle other code up the call stack).

**Comment by arcturus2** [comment]
Hi @serpens24 i just had this happen to me. I know that we discussed about the renderer exception not visually effecting anything. For me though I got an NPE a few seconds after the renderer exception. After this NPE I was not able to interact with the graph so there are actually real issues with this.

Note that the reproducer for me was just trying to open a large graph; I didn't get to the point of opening the Analytic View or any other view.

@Nova-2119 FYI.

The details of the NPE are:

```
java.lang.NullPointerException
    at au.gov.tac.constellation.visual.opengl.renderer.batcher.LineBatcher.drawBatch(LineBatcher.java:325)
    at au.gov.tac.constellation.visual.opengl.renderer.GraphRenderable.display(GraphRenderable.java:504)
    at au.gov.tac.constellation.visual.opengl.renderer.GLRenderer.lambda$display$3(GLRenderer.java:170)
...
...
Caused: com.jogamp.opengl.GLException: Caught NullpointerException: null on thread AWT-EventQueue-1
...
...
Caused: java.lang.RuntimeException
    at com.jogamp.common.util.awt.AWTEDTExecutor.invoke(AWTEDTExecutor.java:58)
...
```

**Comment by Nova-2119** [comment]
Will start looking at this today.
After today I have a few days off so @serpens24 @canis-majoris-42 I will probably reach out to you both on Thursday or Friday next week to discuss this ticket, as mentioned in the last meeting.

**Comment by Nova-2119** [comment]
@arcturus2  I think that error may be a separate error given your description and the stack trace. Are you able to recreate it?

**Comment by Nova-2119** [comment]
Okay so I think I have fixed it.
After some digging I found that in the [NodeLabelBatcher.createBatch](https://github.com/constellation-app/constellation/blob/master/CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java) method sometimes the topLabelFloat/bottomLabelFloats objects were changing in size prior to the between line 159 and 161 being run which was causing the issue. This would lead to an incorrect buffer being created and then the issue with sizes not matching up.
I was unable to pinpoint exactly what code was being run concurrently but figured that a concurrently running thread modifying the same member variable is likely to blame (based on this observation and how randomly the error occurs).
Looking closer at the code I could see no reason for those variables to be member variables as they are recreated every time they are used (and private) so I reformatted them to no longer be member variables.

Since then I have been unable to replicate the problem. I would like other to test and try to replicate the problem if possible.
